pkg/scalers/newrelic_scaler.go

```diff
@@ -0,0 +1,195 @@
+	nrQueryKey = "nrQueryKey"
+	nrRegion   = "nrRegion"
+	metricName = "metricName"
+	nrLogLevel = "nrLogLevel"
+	nrql       = "nrql"
+	threshold  = "threshold"
+)
+
+type newrelicScaler struct {
+	metadata *newrelicMetadata
+	nrClient *newrelic.NewRelic
+}
+
+type newrelicMetadata struct {
+	nrAccount  int
+	nrRegion   string
+	nrLogLevel string
+	nrQueryKey string
+	metricName string
+	nrql       string
+	threshold  int
+
+	scalerIndex int
+}
+
+var newrelicLog = logf.Log.WithName("newrelic_scaler")
+
+func NewNewRelicScaler(config *ScalerConfig) (Scaler, error) {
+	meta, err := parseNewRelicMetadata(config)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing prometheus metadata: %s", err)
+	}
+
+	nrClient, err := newrelic.New(
+		newrelic.ConfigPersonalAPIKey(meta.nrQueryKey),
+		newrelic.ConfigRegion(meta.nrRegion),
+		newrelic.ConfigLogLevel(meta.nrLogLevel))
+
+	if err != nil {
+		log.Fatal("error initializing client:", err)
+	}
+
+	return &newrelicScaler{
+		metadata: meta,
+		nrClient: nrClient}, nil
+}
+
+func parseNewRelicMetadata(config *ScalerConfig) (*newrelicMetadata, error) {
+	meta := newrelicMetadata{}
+	var err error
+	if val, ok := config.TriggerMetadata[nrAccount]; ok && val != "" {
+		t, err := strconv.Atoi(val)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing %s: %s", nrAccount, err)
+		}
+		meta.nrAccount = t
+	} else {
+		return nil, fmt.Errorf("no %s given >%s<", nrAccount, val)
+	}
+
+	meta.nrQueryKey, err = GetFromAuthOrMeta(config, nrQueryKey)
+
+	if err != nil {
+		return nil, fmt.Errorf("no %s given", nrQueryKey)
+	}
+
+	if val, ok := config.TriggerMetadata[metricName]; ok && val != "" {
+		meta.metricName = val
+	} else {
+		return nil, fmt.Errorf("no %s given", metricName)
+	}
+
+	if val, ok := config.TriggerMetadata[nrql]; ok && val != "" {
+		meta.nrql = val
+	} else {
+		return nil, fmt.Errorf("no %s given", nrql)
+	}
+
+	meta.nrQueryKey, err = GetFromAuthOrMeta(config, nrRegion)
+
+	if err != nil {
+		meta.nrRegion = "US"
+		newrelicLog.Info("Using default \"US\" region")
+	}
+
+	meta.nrQueryKey, err = GetFromAuthOrMeta(config, nrLogLevel)
+	if err != nil {
+		meta.nrLogLevel = "INFO"
+	}
+
+	if val, ok := config.TriggerMetadata[threshold]; ok && val != "" {
+		t, err := strconv.Atoi(val)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing %s: %s", threshold, err)
+		}
+
+		meta.threshold = t
+	}
+
+	meta.scalerIndex = config.ScalerIndex
+
+	return &meta, nil
+}
+
+func (s *newrelicScaler) IsActive(ctx context.Context) (bool, error) {
+	val, err := s.ExecuteNewRelicQuery(ctx)
+	if err != nil {
+		newrelicLog.Error(err, "error executing newrelic query")
+		return false, err
+	}
+	return val > 0, nil
+}
+
+func (s *newrelicScaler) Close(context.Context) error {
+	return nil
+}
+
+func (s *newrelicScaler) ExecuteNewRelicQuery(ctx context.Context) (float64, error) {
+
+	nrdbQuery := nrdb.NRQL(s.metadata.nrql)
+	resp, err := s.nrClient.Nrdb.Query(s.metadata.nrAccount, nrdbQuery)
```

Comment:
It would appear that Query does in fact does a QueryWithContext. See: https://github.com/newrelic/newrelic-client-go/blob/2932f5d6275d9017fd1ce5764d2de258575dd187/pkg/nrdb/nrdb_query.go#L6-L8. It eliminates needing to create the context.Context() object within the client, so might be useful to keep this as Query.

Reply:
I change it to use QueryWithContext to use the scaler context instead of the Background context that is used by default on Nrdb.Query , again I'm not a Go expert, so any advice is more than welcome here!